### PR TITLE
move restartmanager out of libcontainerd

### PR DESCRIPTION
**Pull request by tonistiigi**
follow-up for the discussion in #27033

### CODE REVIEW

**Comment by LK4D4** [comment]
@tonistiigi compilation failure on windows.

**Comment by tonistiigi** [comment]
@jhowardmsft  Linux CI passed. Can you take care of the windows side of this change?

@anusha-ragunathan  Can you review the plugin changes as I have limited ways of testing it. One of the issues seems to be that manager and plugin are defined in separate packages. This results in unsafe access to internal properties exposed in public API. If plugin should be a package for the spec then it should have a wrapper in manager. I removed the manager lock because I didn't see it being used for anything. It should definitely be added back so that it protects enable/disable/shutdown ops. The locks I added only protect against data races.

cc @mlaventure @WeiZhang555 for review

**Comment by jhowardmsft** [comment]
Depends on timeline. I'm not quite sure I understand these changes quite yet....

**Comment by tonistiigi** [comment]
@jhowardmsft The restart state is removed, daemon starts container again after receiving exit state. Normally it should mean just removing all restart related code from libcontainerd, any time container exits just send the exit state.

File: daemon/start.go [hunk 1]
```diff
@@ -78,23 +77,23 @@ func (daemon *Daemon) ContainerStart(name string, hostConfig *containertypes.Hos
...
 // container needs, such as storage and networking, as well as links
 // between containers. The container is left waiting for a signal to
 // begin running.
-func (daemon *Daemon) containerStart(container *container.Container, checkpoint string) (err error) {
+func (daemon *Daemon) containerStart(container *container.Container, checkpoint string, resetRestartManager bool) (err error) {
 	container.Lock()
 	defer container.Unlock()
 
-	if container.Running {
+	if resetRestartManager && container.Running {
```
Comment (mlaventure) [inline daemon/start.go]: should this be `||` ?

Comment (tonistiigi) [inline daemon/start.go]: No. Only fresh starts should return `nil`. During restarts `container.Running` is still true but we don't want to return out here. It should probably have a comment. Ideally this check should not exist here at all or instead return a typed error but didn't want to refactor this too much.

**Comment by jhowardmsft** [comment]
@tonistiigi https://github.com/docker/docker/commit/e952b104007521859c817902d73185b0867c83da. Just a tidy up. Have run Windows CI locally, not not with race detection.

**Comment by tonistiigi** [comment]
@jhowardmsft I think the change I did earlier was exaclty the same. CI passed as well.

File: daemon/monitor.go [hunk 2]
```diff
@@ -31,42 +33,53 @@ func (daemon *Daemon) StateChanged(id string, e libcontainerd.StateInfo) error {
...
+				if err != nil {
+					c.SetStopped(platformConstructExitStatus(e))
+					defer autoRemove()
+					if err != restartmanager.ErrRestartCanceled {
+						logrus.Errorf("restartmanger wait error: %+v", err)
+					}
+				}
+			}()
 		}
-		return daemon.postRunProcessing(c, e)
```
Comment (jhowardmsft) [inline daemon/monitor.go]: @tonistiigi This removal will break Windows servicing... @brian-young

Comment (jhowardmsft) [inline daemon/monitor.go]: @brian-young. I don't think that will have any impact, the servicing operation will happen on the first exit, but not on subsequent ones.

Comment (jhowardmsft) [inline daemon/monitor.go]: @swernli too ^^. Can you have a glance over this too pls.

**Comment by anusha-ragunathan** [comment]
@tonistiigi : docker_cli_daemon_experimental_test.go covers restart tests for plugins. plugin is separated from manager, since it primarily implements the manifest specification and its external facing. Manager on the other hand is the entity known to the daemon and is an internal detail. Yes, I will do a followup PR to update manager with wrapper calls to plugin.

File: plugin/manager_linux.go [hunk 3]
```diff
@@ -26,20 +23,15 @@ func (pm *Manager) enable(p *v2.Plugin, force bool) error {
...
-
-	p.RestartManager = restartmanager.New(container.RestartPolicy{Name: "always"}, 0)
-	if err := pm.containerdClient.Create(p.GetID(), "", "", specs.Spec(*spec), libcontainerd.WithRestartManager(p.RestartManager)); err != nil {
-		if err := p.RestartManager.Cancel(); err != nil {
-			logrus.Errorf("enable: restartManager.Cancel failed due to %v", err)
-		}
+	p.Lock()
+	p.Restart = true
+	p.Unlock()
+	if err := pm.containerdClient.Create(p.GetID(), "", "", specs.Spec(*spec)); err != nil {
```
Comment (anusha-ragunathan) [inline plugin/manager_linux.go]: Set p.Restart to false on err.

File: plugin/manager_linux.go [hunk 4]
```diff
@@ -50,49 +42,35 @@ func (pm *Manager) enable(p *v2.Plugin, force bool) error {
...
 }
 
 func (pm *Manager) disable(p *v2.Plugin) error {
 	if !p.IsEnabled() {
 		return fmt.Errorf("plugin %s is already disabled", p.Name())
 	}
-	if err := p.RestartManager.Cancel(); err != nil {
-		logrus.Error(err)
-	}
+	p.Restart = false
```
Comment (anusha-ragunathan) [inline plugin/manager_linux.go]: protect with p.Lock()

File: plugin/store/store_experimental.go [hunk 5]
```diff
@@ -117,7 +117,6 @@ func (ps *Store) Remove(p *v2.Plugin) {
 	delete(ps.plugins, p.GetID())
 	delete(ps.nameToID, p.Name())
 	ps.updatePluginDB()
-	p.RemoveFromDisk()
```
Comment (anusha-ragunathan) [inline plugin/store/store_experimental.go]: Please update function comment to remove "and disk"

**Comment by tonistiigi** [comment]
@anusha-ragunathan updated

**Comment by anusha-ragunathan** [comment]
Plugin changes LGTM.

**Comment by jhowardmsft** [comment]
Restart manager changes LGTM from the Windows side. Haven't reviewed the Linux specifics.

File: daemon/monitor.go (hunk 2 above)
Comment (brian-young) [inline daemon/monitor.go]: It's fine to call `postRunProcessing` multiple times; if no action is necessary, it should be a no-op. Thanks!

**Comment by mlaventure** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**